File: golang/671.second-minimum-node-in-a-binary-tree.go
func findSecondMinimumValue(root *TreeNode) int {
	inf := -1
	first, second := inf, inf

	q := make([]*TreeNode, 0)
	q = append(q, root)

	for len(q) > 0 {
		node := q[0]
		q = q[1:]

		if node.Val < first || first == inf {
			second = first
			first = node.Val
		} else if (node.Val < second || second == inf) && node.Val != first {
			second = node.Val
		}

		if node.Left != nil {
			q = append(q, node.Left)
		}

		if node.Right != nil {
			q = append(q, node.Right)
		}
	}

	return second
}

File: python/671.second-minimum-node-in-a-binary-tree.py
# Definition for a binary tree node.
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findSecondMinimumValue(self, root: Optional[TreeNode]) -> int:
        if not root:
            return -1

        inf = 2**32 - 1
        first, second = inf, inf

        q = deque([root])
        while q:

            node = q.popleft()
            if first > node.val:
                first, second = node.val, first
            elif second > node.val and node.val != first:
                second = node.val

            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

        if second == inf:
            return -1
        return second

